core/commands/storage/challenge/dc_challenge.go

```diff
@@ -0,0 +1,232 @@
+		if err != nil {
+			fmt.Printf("unable to submit challenge job result for peer id {%s}: %v\n", peerId, err)
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		fmt.Printf("grpc error of submit challenge job result for peer id {%s} : %v\n", peerId, err)
+		return nil, err
+	}
+	if result.Code != guardpb.ResponseCode_SUCCESS {
+		fmt.Printf("submit challenge job response error for peer id {%s}, response code {%d}\n", peerId, result.Code)
+		return nil, fmt.Errorf("submit challenge job response error for peer id {%s}, response code {%d}", peerId, result.Code)
+	}
+	return result, nil
+}
+
+func doChallenge(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment, question *Question, requestChan <-chan struct{}, resultChan chan<- *guardpb.ShardChallengeResult) {
+	defer wg.Done()
+	wg.Add(1)
+	errChan := make(chan error)
+	challengeResult := &guardpb.ShardChallengeResult{
+		HostPid:   question.HostPid,
+		FileHash:  question.FileHash,
+		ShardHash: question.ShardHash,
+		Nonce:     question.Nonce,
+	}
+	go func() {
+		err := backoff.Retry(func() error {
+			chunkIndex := strconv.Itoa(int(question.ChunkIndex))
+			storageChallengeRes, err := ReqChallengeStorage(req, res, env, question.HostPid, question.FileHash, question.ShardHash, chunkIndex, question.Nonce)
+			if err != nil {
+				return err
+			}
+			challengeResult.IsTimeout = false
+			challengeResult.Result = storageChallengeRes.Answer
+			return nil
+		}, challengeHostBo)
+		errChan <- err
+	}()
+	select {
+	case err := <-errChan:
+		if err != nil {
+			fmt.Println("failed to challenge shard", zap.String("shard hash", question.ShardHash), zap.String("host pid", question.HostPid), zap.Error(err))
+		}
+		resultChan <- challengeResult
+		<-requestChan
+		return
+	case <-time.After(hostChallengeTimeout):
+		challengeResult.IsTimeout = true
+		resultChan <- challengeResult
+		<-requestChan
+		fmt.Println("challenge shard timeout", zap.String("shard hash", question.ShardHash), zap.String("host pid", question.HostPid))
+		return
+	}
+}
+
+func checkChallengeResults(questionNum int, resultChan <-chan *guardpb.ShardChallengeResult) []*guardpb.ShardChallengeResult {
+	challengeResults := make([]*guardpb.ShardChallengeResult, 0)
+	ticker := time.NewTicker(60 * time.Second)
+	defer ticker.Stop()
+	for range ticker.C {
+		if questionNum == len(resultChan) {
+			for result := range resultChan {
+				challengeResults = append(challengeResults, result)
+			}
+			break
+		}
+	}
+	return challengeResults
+}
+
+func requestQuestions(questionUrl string) ([]*Question, error) {
+	resp, err := http.Get(questionUrl)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	rawData, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("receive wrong status code %d response: %s", resp.StatusCode, string(rawData))
+	}
+
+	var questions []*Question
```

Comment:
I did some code change today for challenge question to add fileHash, so that you can use proto.decentralQuestions to unmarshall the file.   And to reduce the bandwidth, I use gzip to zip the file. The PR is here. 

https://github.com/TRON-US/go-btfs-guard/pull/507/

Reply:
ok, code changed to read gzip file.

Reply:
thanks, code changed to read and unmarshal gzip file.